src/runtime/backend_loader.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ *
+ * Copyright (c) 2021 Aksel Alpay
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "hipSYCL/runtime/backend_loader.hpp"
+#include "hipSYCL/common/debug.hpp"
+#include "hipSYCL/common/config.hpp"
+
+#include <cassert>
+#include <filesystem>
+#include <dlfcn.h>
+
+namespace {
+
+void close_plugin(void *handle) {
+  int err = dlclose(handle);
+
+  if (err != 0) {
+    HIPSYCL_DEBUG_ERROR << "backend_loader: dlclose() failed" << std::endl;
+  }
+}
+
+bool load_plugin(const std::string &filename, void *&handle_out,
+                 std::string &backend_name_out) {
+  void *handle = dlopen(filename.c_str(), RTLD_NOW);
+
+  if (!handle) {
```

Comment:
Again (but here little less of just preference and more of a really useful thing), if you invert the `if` you can do: 
`if(void *handle = ..)` and the handle will always be valid in the scope, it is present in..
helps when thinking about, whether something is valid deep inside a branch :)

Reply:
Good idea, I have applied this.